connect/runtime/src/main/java/org/apache/kafka/connect/runtime/ConfigHash.java

```diff
@@ -0,0 +1,245 @@
+                    e
+            );
+        }
+
+        int hash = Utils.murmur2(serialized);
+        return new ConfigHash(hash);
+    }
+
+    /**
+     * Read and parse the hash from the headers of a REST request.
+     *
+     * @param map the map to read the hash field from; may be null
+     * @param field the field to read and parse; may be null
+     *
+     * @return the parsed hash; never null, but may be {@link #NO_HASH} if
+     * the map is null or the field is not present in the map
+     *
+     * @throws ConnectException if the expected field is present for the hash,
+     * but it cannot be parsed as a 32-bit signed integer
+     */
+    public static ConfigHash fromMap(Map<String, ?> map, String field) {
+        if (map == null)
+            return NO_HASH;
+
+        Object rawHash = map.get(field);
+        if (rawHash == null)
+            return NO_HASH;
+
+        int hash = ConnectUtils.intValue(rawHash);
+        return new ConfigHash(hash);
+    }
+
+    /**
+     * Determine whether tasks should be restarted based on a previously-stored
+     * hash, and the hash for a connector config that was used to generate new task configs
+     *
+     * @param previous the previously-stored config hash for the connector
+     * @param current the hash of the connector config which led to newly-generated
+     *                task configs
+     *
+     * @return whether a restart of the connector's tasks should be forced, possibly to
+     * pick up runtime-controlled configuration changes that would otherwise be dropped
+     *
+     * @see <a href="https://issues.apache.org/jira/browse/KAFKA-9228">KAFKA-9228</a>
+     */
+    public static boolean shouldUpdateTasks(ConfigHash previous, ConfigHash current) {
+        if (previous == null || current == null)
+            return false;
+
+        return previous.exists() && !previous.matches(current);
+    }
+
+    /**
+     * Insert this hash (if it {@link #exists() exists}) into a {@link Struct} with the desired field name.
+     *
+     * @param struct the struct to add the hash to; may be null, in which case
+     *               this method becomes a no-op
+     * @param field the name of the field to add the hash under; may be null,
+     *              in which case this method becomes a no-op
+     */
+    public void addToStruct(Struct struct, String field) {
+        if (hash == null || struct == null || field == null)
+            return;
+        struct.put(field, hash);
+    }
+
+    /**
+     * Add this hash (if it {@link #exists() exists}) to a map of HTTP headers
+     *
+     * @param headers the headers map to add this hash to; may be null, in which case
+     *                this method becomes a no-op
+     */
+    public void addToHeaders(Map<String, String> headers) {
+        if (headers == null || !exists())
+            return;
+
+        headers.put(CONNECTOR_CONFIG_HASH_HEADER, Integer.toString(hash));
+    }
```

Comment:
These are the two non-testing methods most prone to leakage, as there's nothing preventing callers from logging the `Struct` or `Map` arguments (or a subset of their contents) after calling this method.

The `addToHeaders` method provides value independent of reduced change for leakage, since it helps fully encapsulate the name of the HTTP header used.

The `addToStruct` method doesn't really provide any additional value, and if it comes across as too paranoid and not worth the hit in readability, I'd be fine with removing it.

Reply:
I think in service of not having a hash getter, these methods are reasonable. I agree that these are the most leak prone, but in order to have an effect elsewhere, the data has to get out somehow. Having this have a dependency on Map and Struct is fine, the alternative of making this class depend on a REST client and the config topic would be insane.